### PR TITLE
initial tests

### PR DESCRIPTION
**Pull request by phamdt**
this should be a good start to tests but would love some additional eyes on this.

notes: 
- i put the devise secret key in an environment variable
- refactored users#create which i believe simplifies things but hopefully the behavior is still as expected 
- some of the changes are just converting 4 space indentation into 2

### CODE REVIEW

**Comment by irubnich** [comment]
This really is fantastic stuff, I want to use these specs as a baseline for merging some of the other PRs in.

File: .ruby-version [hunk 1]
```diff
@@ -0,0 +1 @@
+1.9.3-p547
```
Comment (phamdt) [inline .ruby-version]: this is the highest 1.9.3 i could go with rbenv

File: app/helpers/users_helper.rb [hunk 2]
```diff
@@ -1,2 +1,12 @@
 module UsersHelper
+  def self.unused_referral_code
```
Comment (phamdt) [inline app/helpers/users_helper.rb]: consolidated

File: spec/controllers/users_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    it 'assigns referral code to cookie if code belongs to user' do
+      expect(User).to receive(:find_by_referral_code).with(@referral_code) do
+        User.new
+      end
+
+      get :new, ref: @referral_code
+      expect(cookies[:h_ref]).to eq(@referral_code)
+    end
+
+    it 'continues request when user agent is facebookexternalhit/1.1' do
```
Comment (phamdt) [inline spec/controllers/users_controller_spec.rb]: positive case

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -1,104 +1,86 @@
...
+
+  def handle_ip
+    # Prevent someone from gaming the site by referring themselves.
+    # Presumably, users are doing this from the same device so block
+    # their ip after their ip appears three times in the database.
+
+    address = request.env['HTTP_X_FORWARDED_FOR']
+    return if address.nil?
+
+    current_ip = IpAddress.find_by_address(address)
```
Comment (phamdt) [inline app/controllers/users_controller.rb]: not an instance var anymore

File: spec/controllers/users_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+      expect(User).to receive(:find_by_email).with(cookies[:h_email]) do
+        User.new
+      end
+
+      get :new
+      expect(response).to redirect_to '/refer-a-friend'
+    end
+
+    it 'assigns referral code to cookie if code belongs to user' do
+      expect(User).to receive(:find_by_referral_code).with(@referral_code) do
```
Comment (phamdt) [inline spec/controllers/users_controller_spec.rb]: .with(@referral_code)

File: config/environment.rb [hunk 6]
```diff
@@ -2,4 +2,6 @@
 require File.expand_path('../application', __FILE__)
 
 # Initialize the rails application
-Prelaunchr::Application.initialize!
\ No newline at end of file
+Prelaunchr::Application.initialize!
+
+Rails.logger = Logger.new(STDOUT)
```
Comment (phamdt) [inline config/environment.rb]: i don't remember why i did this

**Comment by irubnich** [comment]
Awesome, merging it in

**Comment by irubnich** [comment]
Thanks so much for working on this on such short notice!

**Comment by ckhatton** [comment]
:cold_sweat: What's going on? How come the project has suddenly kicked into life again?

**Comment by irubnich** [comment]
This is totally a rogue operation by me.

**Comment by phamdt** [comment]
+1 for rogueing

**Comment by ckhatton** [comment]
+1 oh roguetastic :+1:

**Pull request merged**